### ISSUE TITLE
CMake's _phony_ rule adds extra 5 secs on Windows

### ISSUE THREAD

**Issue by malkia** [flutter/flutter#93745]
## Use case

```cmd
flutter build windows --release -v
```

Shows extra 5 seconds spend somewhere. Loading the .sln file, using the "Build Logging" 
extensions, shows 5 seconds spend everytime the app is build (Ctrl+Shift+B) from Visual Studio.

This led to the discovery it's due to the `_phony_` target that makes it rebuilt everytime.

Now, while `hot-reload` is great and one of the best selling thing about flutter, a slow initial
build time, or slow in general build time if you work from the C++ side is not great.

Removing the `_phony_` target from `CMakeLists.txt` made the problem go away
(from 11secs -> 6secs compile), but not sure why it was put in the first place, or 
whether it's safe to do so.

## Proposal

Look into whether it'll be possible to shed these 5 secs (50% of the time right now, spent
on my machine to `flutter build windows --release` . 

[with_phony.txt](https://github.com/flutter/flutter/files/7550758/with_phony.txt)
[without_phony.txt](https://github.com/flutter/flutter/files/7550759/without_phony.txt)

**Comment by malkia** [comment]
```sh
[√] Flutter (Channel beta, 2.8.0-3.1.pre, on Microsoft Windows [Version 10.0.19043.1348], locale en-US)
    • Flutter version 2.8.0-3.1.pre at d:\p\flu\flutter
    • Upstream repository https://github.com/flutter/flutter
    • Framework revision 3dab555040 (4 days ago), 2021-11-12 11:50:00 -0800
    • Engine revision 09f1520e8b
    • Dart version 2.15.0 (build 2.15.0-268.8.beta)

[!] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at d:\android\sdk
    X cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    X Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Professional 2019 16.11.6)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional
    • Visual Studio Professional 2019 version 16.11.31829.152
    • Windows 10 SDK version 10.0.22000.0

[!] Android Studio
    • Android Studio at C:\Users\dstanev\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6514223
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    X android-studio-dir = C:\Users\dstanev\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6514223
    X Android Studio not found at C:\Users\dstanev\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6514223
    • Try updating or re-installing Android Studio.
    • Consider removing your android-studio-dir setting by running:
      flutter config --android-studio-dir=

[√] VS Code, 64-bit edition (version 1.62.2)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 3.28.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19043.1348]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 96.0.4664.45
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 95.0.1020.44

! Doctor found issues in 2 categories.```

**Comment by malkia** [comment]
Build Logging info (top is before - e.g. with_phony.txt), and then later is `without_phony.txt` - it can be seen that `flutter_assemble` would take ~5secs there while only 0.3secs without

![image](https://user-images.githubusercontent.com/51738/142088345-c72c5c06-2018-49e9-aae8-258d1b8647b0.png)
![image](https://user-images.githubusercontent.com/51738/142088431-89ab8543-b993-4318-b128-8d1c46a4434e.png)

The above can be obtained with https://marketplace.visualstudio.com/items?itemName=VisualStudioProductTeam.ProjectSystemTools (IDE extension, first screenshot) and https://msbuildlog.com/ (The external viewer, 2nd screenshot)

**Comment by stuartmorgan** [comment]
> but not sure why it was put in the first place

For the exactly the reason it says in the comment right above it explaining why the `_phony_` is there.

> or whether it's safe to do so.

No, it's not. (Unsurprisingly, I did not go out of my way to explicitly force that step to every build just to make the build slower.)

> it can be seen that `flutter_assemble` would take ~5secs there while only 0.3secs without

Yes, not running something is generally much faster than running it, so that's not surprising. Similarly, you could speed up your build by not actually recompiling your C++ code. But presumably you want changes to be reflected in your build.

**Comment by stuartmorgan** [comment]
There are two possible approaches to improving this:
- add the ability for `flutter` to generate complete input/output lists for the assemble step, and wire that into the CMake build instead of using `phony`. IIRC there's an issue filed for that, and it was quite difficult to get working.
- Figure out why that step takes 5 seconds (I'm assuming you are not changing anything but C++ code in your tests?). Is it failing to no-op internally when it should be? Is that overhead of just launching the script? Something else?

**Comment by christopherfujino** [comment]
@malkia to clarify, are you saying the flutter assemble target adds 5 seconds when there is no new work to be done? I don't see how we could speed up initial build, because we would need to run flutter assemble at least once.

**Comment by stuartmorgan** [comment]
> are you saying the flutter assemble target adds 5 seconds when there is no new work to be done?

Yes, that target is forced to run every time because we don't have a way to express the inputs and outputs, so we can't let CMake determine when it actually needs to run.

**Comment by christopherfujino** [comment]
> > are you saying the flutter assemble target adds 5 seconds when there is no new work to be done?
> 
> Yes, that target is forced to run every time because we don't have a way to express the inputs and outputs, so we can't let CMake determine when it actually needs to run.

Yeah, I thought maybe that 5 seconds was the actual build time that only happens the first time, but I can see even if there's no work to be done, `flutter assemble` runs in a non-trivial amount of time (3 seconds for me).
